all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
+* Use external configuration by default. Users need to opt-in to this for compatibility and clarity reasons (but it is
+  of course easy).
+
+## Specification
+
+### Values and File Format
+
+* The format is a JSON object with the following fields:
+  * `profiles` - Object with keys as profile names. The default profile should be named `default`.
+    * `<name>` - Profile object with configuration values. Profile names are case-insensitive. 💭 Why? They are needed
+      for environment variables which are case insensitive. It is a validation error to have a config file with two
+      separately-cased profile names that are equal case-insensitively.
+      * `address` - Client address, aka gRPC "host:port".
+      * `namespace` - Client namespace.
+      * `apiKey` - Client API key.
+      * `tls` - A boolean (true is same as empty object) _or_ an object. Note the default for this value is dependent
+        upon the client and may change. 💭 Why? We regret not making TLS the default in the past. If TLS is an object,
+        it can have the following possible fields:
+        * `clientCertPath` - File path to mTLS client cert. Mutually exclusive with `clientCertData`.
+        * `clientCertData` - Cert data. Mutually exclusive with `clientCertPath`.
+        * `clientKeyPath` - File path to mTLS client key. Mutually exclusive with `clientKeyData`.
+        * `clientKeyData` - Key data. Mutually exclusive with `clientKeyPath`.
+        * `serverCaCertPath` - File path to server CA cert. Mutually exclusive with `serverCaCertData`.
+        * `serverCaCertData` - CA cert data. Mutually exclusive with `serverCaCertPath`.
+        * `serverName` - Override SNI name when connecting to server.
+        * `disableHostVerification` - Boolean to disable verifying TLS server host. May not be available in all Temporal
+          clients. 💭 Why? Not all SDKs offer this feature today.
+      * `codec` - Remote codec information to use for all encoding/decoding of payloads. May not be available in all
+        Temporal clients. 💭 Why? Not all SDKs allow use of a remote codec today.
+        * `endpoint` - Endpoint to the remote codec.
+        * `auth` - Authorization header for the remote codec.
+      * `grpcMeta` - Object representing HTTP headers. Values must be strings for now (can discuss arrays later).
+        ❓ Should we call this `headers`? We do in some cases and not others.
+  * 💭 Why JSON?
+    * Because YAML and TOML and others require a dependency in our SDKs. AWS for example has a homegrown INI-ish
+      implementation they have had to build in to every language.
```

Comment:
It might be worth seeing if all of our SDKs actually already include some YAML/TOML dep transitively. If we do, I'd say let's go for one of those, but, JSON is also fine.

Reply:
I can say I know multiple don't. In fact, it's even worse, not all SDKs even have a JSON parser. The only thing they all have is protobuf (which includes protobuf JSON). Here's my understanding:

* Go - Proto yes, JSON stdlib, YAML yes transitively (stupid testify), TOML no
* Java - Proto yes, JSON no (in serviceclient where we need it), YAML no, TOML no
* TypeScript - Proto yes, JSON stdlib, YAML no, TOML no
* Python - Proto yes, JSON stdlib, YAML doubtful, TOML doubtful
* .NET - Proto yes, JSON stdlib, YAML no, TOML no

Sure we can have Core bake in these things for the last 3, but Java is still hurt here, and I think proto JSON works well here.

Reply:
As I suggested above, we should consider vendoring also, so it would be up to the SDK/CLI whether they wish to vendor or use a dependency.

To make it concrete, here's an example of what it might look like in TOML. Not supporting multiple profiles, since whether or not we want that is a separate question that we're discussing. As I've said above, I think there's a strong argument for single-profile per file.

```toml
[client]
address = "localhost:7233"
namespace = "default"
api_key = "<my-key>"
codec = { auth = "<my-headers>", endpoint = "https://my-codec-server.com" }

[client.tls]
client_cert_path = "/path/to/certs/client/cert.pem"
client_key_path = "/path/to/certs/client/key.pem"
server_ca_cert_path = "/path/to/certs/ca/cert.pem"

[client.grpc_meta]
my_grpc_header_1 = "some-value"
my_grpc_header_2 = "some-value"
```